bin/Macronify/MacroniCodeGenVisitorMixin.hpp

```diff
@@ -0,0 +1,124 @@
+        std::int64_t lock_level = 0;
+
+        mlir::Operation *Visit(const clang::Stmt *stmt) {
+            if (clang::isa<clang::ImplicitValueInitExpr,
+                clang::ImplicitCastExpr>(stmt)) {
+                // Don't visit implicit expressions
+                return VisitNonMacro(stmt);
+            }
+
+            // Find the lowest macro that covers this statement, if any
+            auto pasta_stmt = ast->Adopt(stmt);
+            auto sub = lowest_unvisited_substitution(pasta_stmt, visited);
+            if (!sub) {
+                // If no substitution covers this statement, visit it normally.
+                return VisitNonMacro(stmt);
+            }
+
+            // Get the substitution's location, name, parameter names, and
+            // whether it is function-like
+            auto loc = StmtVisitor::meta_location(*sub);
+            auto name_tok = sub->NameOrOperator();
+            auto macro_name = (name_tok
+                               ? name_tok->Data()
+                               : "<a nameless macro>");
+            auto function_like = is_function_like(*sub);
+            auto parameter_names = get_parameter_names(*sub);
+
+            // We call `make_maybe_value_yield_region` here because a macro may
+            // not expand to an expression
+            auto [region, return_type] = make_maybe_value_yield_region(stmt);
+
+            // Check if the macro is an expansion or a parameter, and return the
+            // appropriate operation
+            if (sub->Kind() == pasta::MacroKind::kExpansion) {
+                return StmtVisitor::template make<macroni::MacroExpansion>(
+                    loc,
+                    builder->getStringAttr(llvm::Twine(macro_name)),
+                    builder->getStrArrayAttr(llvm::ArrayRef(parameter_names)),
+                    builder->getBoolAttr(function_like),
+                    return_type,
+                    std::move(region)
+                );
+            } else {
+                return StmtVisitor::template make<macroni::MacroParameter>(
+                    loc,
+                    builder->getStringAttr(llvm::Twine(macro_name)),
+                    return_type,
+                    std::move(region)
+                );
+            }
+        }
+
+        mlir::Operation *VisitNonMacro(const clang::Stmt *stmt) {
+            auto op = StmtVisitor::Visit(stmt);
+            auto call_op = mlir::dyn_cast<vast::hl::CallOp>(op);
+            if (!call_op) {
+                return op;
+            }
+
+            auto name = call_op.getCalleeAttr().getValue();
+            if ("rcu_read_lock" == name) {
+                call_op.getOperation()->setAttr(
+                    "lock_level",
+                    builder->getI64IntegerAttr(lock_level));
+                lock_level++;
+            } else if ("rcu_read_unlock" == name) {
+                lock_level--;
+                call_op.getOperation()->setAttr(
```

Comment:
Still a copy & paste.

Reply:
What exactly is the copy paste code smell you are seeing? How I am setting the `"lock_level"` attribute?

Reply:
Yup, unless I am missing something?

Reply:
No, just confirming